app/clients/zendesk_sell.py

```diff
@@ -141,12 +151,55 @@ def upsert_lead(self, contact: ContactRequest) -> int:
 
         return resp.status_code
 
-    def upsert_contact(self, user: User) -> (Optional[int], bool):
+    def search_lead_id(self, user: User) -> Optional[str]:
+        resp, e = self._send_request(method='GET',
+                                     relative_url=f'/v2/leads?email={user.email_address}')
+        if e:
+            current_app.logger.warning('Failed to create zendesk sell contact')
+            return None
+
+        try:
+            # default to the first lead as we try to perform lead upsert
+            # There SHOULDN'T be any case where there is more than 1 entry
+            resp_dict = resp.json()
+            return resp_dict["items"][0]['data']['id']
+        except (json.JSONDecodeError, KeyError):
+            current_app.logger.warning(f'Invalid response: {resp.text}')
+            return None
+
+    def convert_lead_to_contact(self, user: User) -> Optional[str]:
+
+        lead_id = self.search_lead_id(user)
+        if not lead_id:
+            return None
+
+        # The API and field definitions are defined here:
+        # https://developers.getbase.com/docs/rest/reference/lead_conversions
+
+        resp, e = self._send_request(method='POST',
+                                     relative_url='/v2/lead_conversions',
+                                     data=json.dumps(ZenDeskSell._generate_lead_conversion_data(lead_id)))
+        if e:
+            current_app.logger.warning('Failed to create zendesk sell contact')
+            return None
+
+        try:
+            resp_dict = resp.json()
+            return resp_dict['data']['individual_id']
+        except (json.JSONDecodeError, KeyError):
+            current_app.logger.warning(f'Invalid response: {resp.text}')
+            return None
+
+    def upsert_contact(self, user: User, contact_id: Optional[str]) -> (Optional[int], bool):
 
         # The API and field definitions are defined here: https://developers.getbase.com/docs/rest/reference/contacts
+        if contact_id:
+            rel_path = f'/v2/contacts/upsert?contact_id={contact_id}'
+        else:
+            rel_path = f'/v2/contacts/upsert?custom_fields[notify_user_id]={str(user.id)}'
```

Comment:
Why do we have 2 ways to upsert a contact?

Reply:
There are now two ways of creating a contact, one is when we convert a lead into a contact. When we initiate this work flow the lead does NOT have a `notify_user_id` populated within the lead entry. Truthfully, we don't know a lead has ever registered/logged into the notify system. When they then create a service, which triggers the conversion, we want to perform the conversion from lead->contact, and then we want to upsert that contact based upon the newly created contact_id with the custom fields that we need. Specifically populate the new contact with the `notify_user_id`

On the other hand, when a user hasn't requested a demo and gone straight into creating a user and registering the service. The contact created will have the `notify_user_id`. We use this as a unique identifier for upserting any additions to the contact. Since we do not save the zendesk sell `contact_id` we can not upsert directly and rely on the `notify_user_id` as the unique identifier